File: 01-arrays_&_hashing/00242-Valid_Anagram/00242-valid_anagram.java
// 242: Valid Anagram
// https://leetcode.com/problems/valid-anagram/

class Solution {
    // SOLUTION
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;

        int n = s.length();
        int counts[] = new int[26];
        
        for (int i=0; i<n; i++) {
            counts[s.charAt(i) - 'a']++;
            counts[t.charAt(i) - 'a']--;
        }

        for (int i=0; i<26; i++)
            if (counts[i] != 0) return false;
        
        return true;
    }

    public static void main(String[] args) {
        Solution o = new Solution();
        
        // INPUT
        String s = "rat";
        String t = "cat";

        // OUTPUT
        var result = o.isAnagram(s, t);
        System.out.println((result ? "true" : "false"));
    }
}

File: 01-arrays_&_hashing/00242-Valid_Anagram/00242-valid_anagram.cpp
// 242: Valid Anagram
// https://leetcode.com/problems/valid-anagram/

#include <iostream>

using namespace std;

class Solution {
public:
    // SOLUTION
    bool isAnagram(string s, string t) {
        if (s.length() != t.length()) return false;

        int n = s.length();
        int counts[26] = {0};
        
        for (int i=0; i<n; i++) {
            counts[s[i] - 'a']++;
            counts[t[i] - 'a']--;
        }
        
        for (int i=0; i<26; i++)
            if (counts[i]) return false;
        
        return true;
    }
};

int main() {
    Solution o;

    // INPUT
    string s = "rat";
    string t = "cat";

    // OUTPUT
    auto result = o.isAnagram(s, t);
    cout << (result ? "true" : "false") << endl;

    return 0;
}
